vendor/k8s.io/kubernetes/plugin/pkg/admission/namespace/lifecycle/admission.go

```diff
@@ -80,26 +82,35 @@ func (l *lifecycle) Admit(a admission.Attributes) (err error) {
 		return nil
 	}
 
-	namespaceObj, exists, err := l.store.Get(&api.Namespace{
-		ObjectMeta: api.ObjectMeta{
-			Name:      a.GetNamespace(),
-			Namespace: "",
-		},
-	})
+	namespaceObj, exists, err := l.store.Get(&api.Namespace{ObjectMeta: api.ObjectMeta{Name: a.GetNamespace()}})
 	if err != nil {
 		return errors.NewInternalError(err)
 	}
 
+	if !exists {
+		// give the cache time to observe a recent namespace creation
+		time.Sleep(50 * time.Millisecond)
+		namespaceObj, exists, err = l.store.Get(&api.Namespace{ObjectMeta: api.ObjectMeta{Name: a.GetNamespace()}})
+		if err != nil {
+			return errors.NewInternalError(err)
+		}
+		if exists {
+			glog.V(4).Infof("found %s in cache after waiting", a.GetNamespace())
+		}
+	}
+
 	// refuse to operate on non-existent namespaces
 	if !exists {
-		// in case of latency in our caches, make a call direct to storage to verify that it truly exists or not
+		// as a last resort, make a call directly to storage
+		// this also benefits from the Sleep() above allowing for propagation in HA storage cases
 		namespaceObj, err = l.client.Core().Namespaces().Get(a.GetNamespace())
```

Comment:
I also would be fine skipping this, using a poller for X milliseconds on the cache, and skipping the "live" check altogether.